Allow SLAssert methods to be called from non SLTest subclass methods

justinseanmartin opened the pull request:
What do you think about this? I was planning on building a library of helper methods to build up the test cases, but unless they subclass SLTest, I can't call the assert methods. This doesn't seem right to me. 

This is my proposed solutions, but I wanted to get some thoughts.

wearhere commented on Sources/Classes/SLTest.m:
```diff
@@ -249,40 +249,27 @@ - (BOOL)runAndReportNumExecuted:(NSUInteger *)numCasesExecuted
-
-                // Only execute the test case if set-up succeeded.
-                if (!caseFailed) {
+    for (NSUInteger i = 0; i < 1000; i++) {
```
![](http://i.imgur.com/5M7e2Kg.gif)

Commit:
Allow SLAssert methods to be called from non SLTest subclass methods

wearhere commented:
A potential concern with letting the assertions be called outside of a test would be that they could be called outside of the current test's exception-handling infrastructure. Would you just leave it up to the user to be aware of/guard against that themselves?

Two alternatives in which you might be interested would be to:
1. define the helpers on the test itself. If the helpers are used by more than one test, you could define them on an [abstract test](http://inkling.github.io/Subliminal/Documentation/Classes/SLTest.html#//api/name/isAbstract). At Inkling, for instance, all tests subclass an "InklingTest" which defines methods related to logging in, downloading books, etc. (Obviously, there's no guarantee that these methods are only called from within test cases either, but I'd say it's more likely.)
2. have helpers defined on other classes return an error/`BOOL` if they succeed or not. You could then have the tests assert that success.

justinseanmartin commented:
Yea, I'd leave it up to the user to be aware of only throwing these exceptions inside of test code. Let me look at our architecture and see if I can refactor to something else that makes sense.

wearhere commented on Sources/Classes/SLTest.m:
```diff
@@ -40,10 +40,10 @@
-}
+@implementation SLTest
+
+static NSString *_lastKnownFilename;
```
Please prefix static variables with two underscores to differentiate them from instance variables.

wearhere commented:
I don't think my objections above are that strong. If you can address my stylistic comment above I'll merge this then.

Commit:
Incorporate PR feedback

Adding a comment to SLTest.h describing when SLAssert should be used
Renaming the static variables to use double underscores

justinseanmartin commented:
Done. I also added a comment to SLTest.h on SLAssertTrue for how to do this.

I think there is already a bug saying that we should expand on test case organization, maybe we could add on our use case to that ticket. 

The current design is to break out a class for each screen of our UI with helper methods. This will allow containing of logic for interacting with a particular screen to a single place.

wearhere commented on Sources/Classes/SLTest.h:
```diff
@@ -404,15 +404,21 @@
+ should only be used from test setup, teardown, or execution methods as well as those
+ methods called from within.
+
+ @warning If you perform operations on the main thread within your test case, you can't
```
I think this is a good point, though I find the phrasing a bit confusing. What do you think about

> @warning If you use an assertion inside a dispatch block--if your test case dispatches to the main queue, for instance--you must wrap the assertion in a try-catch block and re-throw the exception it generates (if any) outside the dispatch block. Otherwise, the tests will abort.

wearhere commented on Sources/Classes/SLTest.h:
```diff
@@ -404,15 +404,21 @@
- 
+ Fails the test case if the specified expression is false. The SLAssert class of methods
+ should only be used from test setup, teardown, or execution methods as well as those
+ methods called from within.
```
Let's move this sentence and the @warning below into a comment right below the pragma, not tied specifically to `SLAssertTrue`.

wearhere commented:
Some small comments on documentation. I agree that a discussion of test case organization is out of scope for this request/more appropriate for the wiki.

justinseanmartin commented:
Incorporated the last 2 pieces of feedback. 

It is unfortunate that the SLAssert documentation doesn't seem to appear in the published documentation anywhere.  It seems that is a limitation of appledoc in general though, so not much you can do about it I guess.

wearhere commented:
> It is unfortunate that the SLAssert documentation doesn't seem to appear in the published documentation anywhere.

Yeah, I don't think `appledoc` handles macros at present. (The latest release might: #71.)

Looks good! I'll merge as soon as Travis finishes up.

wearhere commented:
:ok_hand:
